util/src/test-support/java/tech/pegasys/artemis/util/executors/MockExecutorService.java

```diff
@@ -0,0 +1,176 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.artemis.util.executors;
+
+import static tech.pegasys.artemis.util.async.FutureUtil.ignoreFuture;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.stream.Collectors;
+
+public class MockExecutorService implements ExecutorService {
+
+  private boolean autoRun = false;
+
+  private final List<ExecutorTask<?>> tasks = new ArrayList<>();
+
+  // Test utility for inspecting executor's futures
+  public List<Future<?>> getFutures() {
+    return tasks.stream().map(ExecutorTask::getFuture).collect(Collectors.toList());
+  }
+
+  public void setAutoRun(final boolean shouldAutoRunTasks) {
+    this.autoRun = shouldAutoRunTasks;
+  }
+
+  public void runPendingFutures() {
+    final List<ExecutorTask<?>> currentTasks = new ArrayList<>(tasks);
+    currentTasks.forEach(ExecutorTask::run);
+  }
+
+  public long getPendingFuturesCount() {
+    return tasks.stream().filter(ExecutorTask::isPending).count();
+  }
+
+  public void runPendingFuturesInSeparateThreads(final ExecutorService executorService) {
+    final List<ExecutorTask<?>> currentTasks = new ArrayList<>(tasks);
+    currentTasks.forEach(task -> executorService.execute(task::run));
+  }
+
+  @Override
+  public void shutdown() {}
+
+  @Override
+  public List<Runnable> shutdownNow() {
+    return Collections.emptyList();
+  }
+
+  @Override
+  public boolean isShutdown() {
+    return false;
+  }
+
+  @Override
+  public boolean isTerminated() {
+    return false;
+  }
+
+  @Override
+  public boolean awaitTermination(final long timeout, final TimeUnit unit)
+      throws InterruptedException {
+    return false;
+  }
+
+  @Override
+  public <T> Future<T> submit(final Callable<T> task) {
+    ExecutorTask<T> execTask = new ExecutorTask<>(task::call);
+    tasks.add(execTask);
+    if (autoRun) {
+      execTask.run();
+    }
+
+    return execTask.getFuture();
+  }
+
+  @Override
+  public <T> Future<T> submit(final Runnable task, final T result) {
```

Comment:
Did you implement these because extending the interface required you to do so? I don't understand why you return the same exact thing that was passed into the argument.

Reply:
This task gets channeled to the submit method [on line 85](https://github.com/PegaSysEng/artemis/pull/1118/files#diff-e19e1b2013701f87903e809a985b551dR85) which adds tracking arounds the tasks.

Reply:
I grabbed these utils from besu btw: https://github.com/hyperledger/besu/blob/6060a79d6618fbeebf6e5e1da7e9f788a213598a/testutil/src/main/java/org/hyperledger/besu/testutil/MockExecutorService.java